server: Node Restart Event records Last Up time

mrtracy opened the pull request:
Augment the "Node Restart" event with an additional information field "LastUp",
which records the last time at which the node was up before it was restarted.
This field can be used with the existing "StartedAt" to determine the
approximate amount of time the node was down.

A new per-store local key is create which contains the last timestamp at which
the store's node was reported to be up. This timestamp is updated (on all stores
of the node) whenever the liveness record is successfully updated.

When a node restarts (and finds previously bootstrapped stores), it reads this
value from all stores, selecting the most recent timestamp from the stores, and
records this time in the "LastUp" field of the Node Restart event.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13382)
<!-- Reviewable:end -->

Commit:
server: Node Restart Event records Last Up time

Augment the "Node Restart" event with an additional information field "LastUp",
which records the last time at which the node was up before it was restarted.
This field can be used with the existing "StartedAt" to determine the
approximate amount of time the node was down.

A new per-store local key is create which contains the last timestamp at which
the store's node was reported to be up. This timestamp is updated (on all stores
of the node) whenever the liveness record is successfully updated.

When a node restarts (and finds previously bootstrapped stores), it reads this
value from all stores, selecting the most recent timestamp from the stores, and
records this time in the "LastUp" field of the Node Restart event.

bdarnell commented:
The approach seems fine to me.

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/server/node.go, line 484 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kc-aWMcxhbH57Yl-EKG:-Kc-aWMcxhbH57Yl-EKH:b-whnvo8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c085d6c8e42405e27517d47e2ea6f62063bf2b09/pkg/server/node.go#L484)):*
> ```Go
> 
> 	// Compute last up time from stores; this is done by checking the last
> 	// up timestamp for every
> ```

s/every/every store./

---

*[pkg/server/node.go, line 768 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kc-ajwCSk0GX4qsZTBE:-Kc-ajwCSk0GX4qsZTBF:b3hgf12) ([raw file](https://github.com/cockroachdb/cockroach/blob/c085d6c8e42405e27517d47e2ea6f62063bf2b09/pkg/server/node.go#L768)):*
> ```Go
> 
> 	logEventType := sql.EventLogNodeRestart
> 	var eventInfo interface{}
> ```

Probably better to make this the latter of the two `struct` types instead of an opaque interface. 

---

*[pkg/storage/node_liveness.go, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kc-b5_3JwYbqVr3mwJA:-Kc-b5_3JwYbqVr3mwJB:b-x5tors) ([raw file](https://github.com/cockroachdb/cockroach/blob/c085d6c8e42405e27517d47e2ea6f62063bf2b09/pkg/storage/node_liveness.go#L87)):*
> ```Go
> type IsLiveCallback func(nodeID roachpb.NodeID)
> 
> type RecordLocalUp func(context.Context, hlc.Timestamp) error
> ```

s/RecordLocalUp/AliveCallback/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-:-Kc-bccAGpd23zfEEm3Y:b-7zgj4k)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/keys/keys.go, line 55 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kc-dOEPwi-xo_YAMSoL:-Kc-dOEQqsvlyfCsNWZg:bdn8gte) ([raw file](https://github.com/cockroachdb/cockroach/blob/c085d6c8e42405e27517d47e2ea6f62063bf2b09/pkg/keys/keys.go#L55)):*
> ```Go
> }
> 
> // StoreLastUpKey
> ```

fill it in

---

*[pkg/storage/store.go, line 1051 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kc-dix7QOjynNM5vrBM:-Kc-dix7QOjynNM5vrBN:bdn8gte) ([raw file](https://github.com/cockroachdb/cockroach/blob/c085d6c8e42405e27517d47e2ea6f62063bf2b09/pkg/storage/store.go#L1051)):*
> ```Go
> }
> 
> // ReadStoreLastUp .
> ```

fill it in

---

*[pkg/storage/store.go, line 1521 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kc-dozQOOs8gRhqEUtv:-Kc-dozQOOs8gRhqEUtw:bdn8gte) ([raw file](https://github.com/cockroachdb/cockroach/blob/c085d6c8e42405e27517d47e2ea6f62063bf2b09/pkg/storage/store.go#L1521)):*
> ```Go
> }
> 
> // WriteLastUpTimestamp .
> ```

fill it in

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382)*
<!-- Sent from Reviewable.io -->

tamird commented:
Missed the last paragraph of your commit message. Seems OK to me.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-:-Kc-dvF9BeVr1CHinpLQ:b-ml6fn7)*
<!-- Sent from Reviewable.io -->

petermattis commented:
For determining the duration of how long a node was down, this seems fine. 

But from a higher perspective, it is unfortunate that we don't have an event when a node goes down. Every other node in the system has a signal (via node liveness) that the down node is down. It seems feasible to design a mechanism for one or more of those nodes to record a node down event. Perhaps they all record the event and we de-duplicate upon reading the event log. Perhaps all of the nodes try to bump the epoch of the down node and only the one that succeeds records the down event. My intuition is that we've done a large chunk of the work for the system to detect a down node, except the final mile of recording an event about it.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-:-Kc-eKv9V6vqMKjfMROw:b-dotnlc)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
I don't believe a specific "Node down" event is necessary for any feature we currently want to provide.  If a node has gone down in the past, it has either come back up or it is still down; those states can be discovered from the rejoin event or the liveness table respectively. This is adequate for all of our planned UI work.

If we do eventually record that event, I think it should not be a "Node Down" event, but rather a "Repair Begin" event recorded by other nodes when they actually take action on a down node and begin repair replicas. Short-term outages will be captured perfectly fine by the rejoin event. "Repair Begin" would be easier to implement (nodes could do it independently) and the mass of individual events would contain meaningful information (specifically, which nodes were involved in repair work).  Those individual events could be consolidated in the UI into an aggregate event.

petermattis commented:
> I don't believe a specific "Node down" event is necessary for any feature we currently want to provide. If a node has gone down in the past, it has either come back up or it is still down; those states can be discovered from the rejoin event or the liveness table respectively. This is adequate for all of our planned UI work.

Let's put aside the question of implementation for a moment. If a node goes down and doesn't come back up, don't we want that event generated and flagged on the graphs? I agree that you can deduce this scenario from the live nodes graph. I disagree that this is adequate. Perhaps I'm missing context here, but it appears your reasoning is being clouded by what you've implemented and what is straightforward to implement.

vivekmenezes commented:
I believe what's being recommended here is all NodeRejoin events along with the current state of the node liveness table should be sufficient to plot down/up events on the graphs over any period of time.

petermattis commented:
> I believe what's being recommended here is all NodeRejoin events along with the current state of the node liveness table should be sufficient to plot down/up events on the graphs over any period of time.

I that what is being suggested? If it is then I'm fine with it. We don't need a NodeDown event, but I definitely want a node going down to be indicated as an overlayed event on the graphs even if it never come back up.

mrtracy commented:
PR is updated and ready for normal approval. Updated the PR title/text to match the commit message.

mrtracy commented:
ping @petermattis @bdarnell @tamird @vivekmenezes; this is falling into that hole where I don't know to whom it should be assigned.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/node_liveness.go, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kc-b5_3JwYbqVr3mwJA:-Kcssgh9NyYyKC-ug8Hq:b-zh0ir7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c085d6c8e42405e27517d47e2ea6f62063bf2b09/pkg/storage/node_liveness.go#L87)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/RecordLocalUp/AliveCallback/
</blockquote></details>

Oh, I didn't see that we already had `IsLiveCallback`. These two things need more distinct names. Maybe call this one `HeartbeatCallback`.

---

*[pkg/storage/node_liveness.go, line 472 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kcss-ae-cVXKkffSU9H:-Kcss-ae-cVXKkffSU9I:b-z42tjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e769c628e663e11ac46ab05674d05f7902b7677/pkg/storage/node_liveness.go#L472)):*
> ```Go
> 		return err
> 	}
> 	aliveAt := nl.clock.Now()
> ```

This should either be a time that only `NodeLiveness` could know (i.e. the time used to compute the new liveness expiration) or remove it from the callback and let the callback call `Now()` if it needs to. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-:-KcsuhCsU-6Xv70qyZ50:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 2 of 8 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/keys/keys.go, line 55 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kc-dOEPwi-xo_YAMSoL:-Kcu6pTO1_n4OHyHxqrT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c085d6c8e42405e27517d47e2ea6f62063bf2b09/pkg/keys/keys.go#L55)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

fill it in
</blockquote></details>

Done.

---

*[pkg/server/node.go, line 484 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kc-aWMcxhbH57Yl-EKG:-KcKhPNZF2Gs4xYOKiu3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c085d6c8e42405e27517d47e2ea6f62063bf2b09/pkg/server/node.go#L484)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/every/every store./
</blockquote></details>

Done.

---

*[pkg/server/node.go, line 768 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kc-ajwCSk0GX4qsZTBE:-KcKhHmSC5wS1M9sSUoq:b-xulo66) ([raw file](https://github.com/cockroachdb/cockroach/blob/c085d6c8e42405e27517d47e2ea6f62063bf2b09/pkg/server/node.go#L768)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Probably better to make this the latter of the two `struct` types instead of an opaque interface. 
</blockquote></details>

Done, it's okay for the Join event to have a LastUp that is ignored and it simplifies this considerably.

---

*[pkg/storage/node_liveness.go, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kc-b5_3JwYbqVr3mwJA:-KcKsR-qfFw59aHQ150t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c085d6c8e42405e27517d47e2ea6f62063bf2b09/pkg/storage/node_liveness.go#L87)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/RecordLocalUp/AliveCallback/
</blockquote></details>

Done.

---

*[pkg/storage/node_liveness.go, line 472 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kcss-ae-cVXKkffSU9H:-Kcu6qRaY9YcQG5EwLxi:b-n4skao) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e769c628e663e11ac46ab05674d05f7902b7677/pkg/storage/node_liveness.go#L472)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This should either be a time that only `NodeLiveness` could know (i.e. the time used to compute the new liveness expiration) or remove it from the callback and let the callback call `Now()` if it needs to. 
</blockquote></details>

Good point, I was originally using the time for the lease extension, but that was incorrect as it is not the actual time the node was reported as alive. I've moved the timestamp generation to the callbacks.

---

*[pkg/storage/store.go, line 1051 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kc-dix7QOjynNM5vrBM:-KcKuyyUuFZYu7uPH6qy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c085d6c8e42405e27517d47e2ea6f62063bf2b09/pkg/storage/store.go#L1051)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

fill it in
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 1521 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kc-dozQOOs8gRhqEUtv:-KcKuyYvFukywQZp9Tlm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c085d6c8e42405e27517d47e2ea6f62063bf2b09/pkg/storage/store.go#L1521)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

fill it in
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/server/server.go, line 682 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-KcuDb-We_BOUNJngTZ6:-KcuDb-We_BOUNJngTZ7:bbmtxv4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1548a01473525485115a4ed1359041b07640921/pkg/server/server.go#L682)):*
> ```Go
> 	// is updated.
> 	s.nodeLiveness.StartHeartbeat(ctx, s.stopper, func(ctx context.Context) error {
> 		time := s.clock.Now()
> ```

"time" is a package, so might be good to call this something else.

---

*[pkg/storage/client_test.go, line 792 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-KcuDp-kSIIXBs4sPtoW:-KcuDp-kSIIXBs4sPtoX:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1548a01473525485115a4ed1359041b07640921/pkg/storage/client_test.go#L792)):*
> ```Go
> 
> 	m.nodeLivenesses[idx].StartHeartbeat(context.Background(), stopper, func(ctx context.Context) error {
> 		time := m.clock.Now()
> ```

ditto

---

*[pkg/storage/client_test.go, line 875 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-KcuDrQ4CF-asVEZNhdL:-KcuDrQ4CF-asVEZNhdM:b-wg78tq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1548a01473525485115a4ed1359041b07640921/pkg/storage/client_test.go#L875)):*
> ```Go
> 
> 	cfg.NodeLiveness.StartHeartbeat(context.Background(), stopper, func(ctx context.Context) error {
> 		time := m.clock.Now()
> ```

and this one

---

*[pkg/storage/node_liveness_test.go, line 207 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-KcuEBO0s6UCpjeL-Nqs:-KcuEBO0s6UCpjeL-Nqt:bbtev7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1548a01473525485115a4ed1359041b07640921/pkg/storage/node_liveness_test.go#L207)):*
> ```Go
> 			uptm, err := storage.ReadStoreLastUp(context.Background(), s.Engine())
> 			if err != nil {
> 				return errors.Errorf("error reading last up time from store %d: %s", i, err)
> ```

errors.Wrapf

---

*[pkg/storage/node_liveness_test.go, line 210 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-KcuEEI4880xa6KLrNUw:-KcuEEI4880xa6KLrNUx:bj4tpex) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1548a01473525485115a4ed1359041b07640921/pkg/storage/node_liveness_test.go#L210)):*
> ```Go
> 			}
> 			if a, e := uptm.WallTime, expected.WallTime; a != e {
> 				return errors.Errorf("store %d last uptime = %d; wanted %d\n", i, a, e)
> ```

what's with the newline?

---

*[pkg/storage/store.go, line 1071 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-KcuETe1JnYo2nYRyDHH:-KcuETe1JnYo2nYRyDHI:b-8aci3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1548a01473525485115a4ed1359041b07640921/pkg/storage/store.go#L1071)):*
> ```Go
> 	var timestamp hlc.Timestamp
> 	ok, err := engine.MVCCGetProto(
> 		ctx, eng, keys.StoreLastUpKey(), hlc.ZeroTimestamp, true, nil, &timestamp)
> ```

why is this style different from the style you use in `WriteLastUpTimestamp`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 3 of 8 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/server/server.go, line 682 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-KcuDb-We_BOUNJngTZ6:-Kcv9v8J1Y0bgHLjs1Fi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1548a01473525485115a4ed1359041b07640921/pkg/server/server.go#L682)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

"time" is a package, so might be good to call this something else.
</blockquote></details>

Done.

---

*[pkg/storage/client_test.go, line 792 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-KcuDp-kSIIXBs4sPtoW:-Kcv9vhv60EGou1ilNNu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1548a01473525485115a4ed1359041b07640921/pkg/storage/client_test.go#L792)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[pkg/storage/client_test.go, line 875 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-KcuDrQ4CF-asVEZNhdL:-Kcv9w6aIFPS_dVIVMA0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1548a01473525485115a4ed1359041b07640921/pkg/storage/client_test.go#L875)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

and this one
</blockquote></details>

Done.

---

*[pkg/storage/node_liveness_test.go, line 207 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-KcuEBO0s6UCpjeL-Nqs:-Kcv79KjpRMoum-Z0iJh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1548a01473525485115a4ed1359041b07640921/pkg/storage/node_liveness_test.go#L207)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

errors.Wrapf
</blockquote></details>

Done.

---

*[pkg/storage/node_liveness_test.go, line 210 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-KcuEEI4880xa6KLrNUw:-Kcv6XJ14uxsBA9Tt7o9:b4pob6e) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1548a01473525485115a4ed1359041b07640921/pkg/storage/node_liveness_test.go#L210)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what's with the newline?
</blockquote></details>

feng shui
(this was a t.Errorf(), changed to the style that you have beaten into me after so many test PRs)

---

*[pkg/storage/store.go, line 1071 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-KcuETe1JnYo2nYRyDHH:-Kcv7A513NJxuUsQ6Qps:b-te8mkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1548a01473525485115a4ed1359041b07640921/pkg/storage/store.go#L1071)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why is this style different from the style you use in `WriteLastUpTimestamp`?
</blockquote></details>

Balancing consistency with the existing API vs the obvious solution. 

The most similar key being written to the store is the StoreIdent, which also uses this pattern (Write a method of store, Read a free function working on the engine). However, in the case of StoreIdent it's necessary because StoreIdent is needed from the engine before the store is created; if that wasn't the case, it's likely both Read and Write would be methods of the Store.

However, they *aren't* both methods of the store, so in my case I was balancing the obvious factoring (Read and Write both methods of stores) vs. consistency with the most similar existing API.

Although, now that I've written it out, I disagree with my own decision; i've changed this to a Store method.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 8 files at r2, 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/server/node.go, line 494 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kcw_dH2zBKJXZF6mlZb:-Kcw_dH2zBKJXZF6mlZc:bb76tpr) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d099765db6b7a35073cf25c45d38077450b3117/pkg/server/node.go#L494)):*
> ```Go
> 		return nil
> 	}); err != nil {
> 		return fmt.Errorf("failed to read last up timestamp from stores: %s", err)
> ```

Wrap

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 5 of 8 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/server/node.go, line 494 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382#-Kcw_dH2zBKJXZF6mlZb:-KcyMargAp2mt_GSBO8M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d099765db6b7a35073cf25c45d38077450b3117/pkg/server/node.go#L494)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Wrap
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13382)*
<!-- Sent from Reviewable.io -->
